contracts/StakeManager.sol

```diff
@@ -0,0 +1,160 @@
+
+import "./interfaces/IRelayHub.sol";
+
+contract StakeManager {
+
+    /// Emitted when a stake or unstakeDelay are initialized or increased
+    event StakeAdded(
+        address indexed registree,
+        address indexed owner,
+        uint256 stake,
+        uint256 unstakeDelay
+    );
+
+    /// Emitted once a stake is scheduled for withdrawal
+    event StakeUnlocked(
+        address indexed registree,
+        address indexed owner,
+        uint256 withdrawBlock
+    );
+
+    /// Emitted when owner withdraws registree funds
+    event StakeWithdrawn(
+        address indexed registree,
+        address indexed owner,
+        uint256 amount
+    );
+
+    /// Emitted when an authorized Relay Hub penalizes a registree
+    event StakePenalized(
+        address indexed registree,
+        address indexed beneficiary,
+        uint256 reward
+    );
+
+    event HubAuthorized(
+        address indexed registree,
+        address indexed relayHub
+    );
+
+    event HubUnauthorized(
+        address indexed registree,
+        address indexed relayHub,
+        uint256 removalBlock
+    );
+
+    /// @param stake - amount of ether staked for this relay
+    /// @param unstakeDelay - number of blocks to elapse before the owner can retrieve the stake after calling 'unlock'
+    /// @param withdrawBlock - first block number 'withdraw' will be callable, or zero if the unlock has not been called
+    /// @param owner - address that receives revenue and manages registree's stake
+    struct StakeInfo {
+        uint256 stake;
+        uint256 unstakeDelay;
+        uint256 withdrawBlock;
+        address payable owner;
+    }
+
+    struct RelayHubInfo {
+        uint256 removalBlock;
+    }
+
+    /// maps registrees to their stakes
+    mapping(address => StakeInfo) public stakes;
+
+    /// maps registrees to their authorized hubs map
+    mapping(address => mapping(address => RelayHubInfo)) public authorizedHubs;
+
+    /// Put a stake for a registree and set its unstake delay.
+    /// If the entry does not exist, it is created, and the caller of this function becomes its owner.
+    /// If the entry already exists, only the owner can call this function.
+    /// @param registree - address that represents a stake entry and controls relay registrations on relay hubs
+    /// @param unstakeDelay - number of blocks to elapse before the owner can retrieve the stake after calling 'unlock'
+    function stakeForAddress(address registree, uint256 unstakeDelay) external payable {
+        require(stakes[registree].owner == address(0) || stakes[registree].owner == msg.sender, "not owner");
+        require(unstakeDelay >= stakes[registree].unstakeDelay, "unstakeDelay cannot be decreased");
+        require(msg.sender != registree, "registree cannot stake for itself");
+        require(stakes[msg.sender].owner == address(0), "sender is a registree itself");
+        stakes[registree].owner = msg.sender;
```

Comment:
maybe its only optimization: but might do 
`if stakes[registree].owner != msg.sender ) stakes[registree].owner = msg.sender`

Reply:
Then I can steal your stake by topping it

Reply:
no. its only "don't write if it doesn't modify". 
in Solidity, doing "this.a=1" costs you a storage, even if a is already 1. wrapping with "if" makes you just pay for condition and read, but avoid writing the same value (though I'm sure future EVM will remove the extra cost of writing same-value)